### PR TITLE
[quant] supports act_order inputs in Matmulnbits and new quantization algorithm "hqq"

### PR DESCRIPTION
**Pull request by wejoncy**
### Description

1. Support quantized GPTQ weight in huggingface like [TheBloke/Llama-2-7B-Chat-GPTQ](https://huggingface.co/TheBloke/Llama-2-7B-Chat-GPTQ)
2. Support Act_order for GPTQ
3. Support [HQQ](https://mobiusml.github.io/hqq_blog/) algorithm to quantize matmul weight and add quant script

### CODE REVIEW

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 1]
```diff
@@ -3333,22 +3333,32 @@ MatMulNBits is a MatMul with weight quantized with N bits(e.g., 2, 3, 4, 5, 6, 7
...
-    uint8 one_bits[(bits & 0x1) * 1 * block_size / 8];  // highest 1 bit for 3, 5, 7 bits quantization
-    uint8 two_bits[(bits & 0x2) * 2 * block_size / 8];  // high 2 bits for 2, 6, 7 bits quantization
-    uint8 four_bits[(bits & 0x4) * 4 * block_size / 8]; // low 4 bits for 4, 5, 6 bits quantization
-  }
-
-Input scales is stored in same type as original type of B(float32, float16) with shape like: [N * n_blocks_per_col]
-Input zero_points is stored as uint8_t. If bits <= 4, two zero points are stored as one unit8_t. If bits > 4, one zero point is stored with one unit8_t. Thus, its shape is:
-  - [(N * n_blocks_per_col + 1) / 2] if bits <=4
-  - [N * n_blocks_per_col] if bits > 4
+It support different packing style, ["default", "gptq", "hqq", "exl2", "awq"], Please note that "exl2", "awq" is not supported yet.
```
Comment (yufenglee) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:3336]: With this attribute, we need to write different kernels/code for different packing mode. I think we need to handle this with offline script instead.

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 2]
```diff
@@ -3333,22 +3333,32 @@ MatMulNBits is a MatMul with weight quantized with N bits(e.g., 2, 3, 4, 5, 6, 7
...
+  Input B is stored as uint32_t with shape: [ceil(N, 32//bits), K]:
+  zero_points has shape uint32_t [N//block_size, K*bits//32]
+  scale has shape float16[N//block_size, K]
+with "hqq", we have:
+  Input B is stored as uint32_t with shape: [ceil(N, 32//bits), K]:
+  zero_points has shape float16 [N//block_size, K]
+  scale has shape [N//block_size, K]
 )DOC";
 
   ONNX_CONTRIB_OPERATOR_SCHEMA(MatMulNBits)
```
Comment (yufenglee) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:3364]: I think we need to add a QuantizeNBitsBlock and DequantizeNBitsBlock for convenience. And then we can fuse the DequantizeNBitsBlock + MatMul into different ops based on the different parameters.

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 3]
```diff
@@ -3325,30 +3325,141 @@
...
+  Input scales is stored in same type as original type of B(float32, float16) with shape like: [N * n_blocks_per_col]
+  Input zero_points is stored as uint8_t. If bits <= 4, two zero points are stored as one unit8_t. If bits > 4, one zero point is stored with one unit8_t. Thus, its shape is:
+    - [(N * n_blocks_per_col + 1) / 2] if bits <=4
+    - [N * n_blocks_per_col] if bits > 4
+
+with "gptq", we have:
+  32 = bits_of(int32)
+  Input B is stored as uint32_t with shape: [ceil(K*bits, 32), N]:
+  zero_points has shape uint32_t [K//block_size, ceil(N*bits, 32)]
+  scale has shape float16[N//block_size, N]
```
Comment (yufenglee) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:3458]: >N [](http://example.com/codeflow?start=26&length=1)

It should be K

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 4]
```diff
@@ -3325,30 +3325,141 @@
...
+with "hqq", we have:
+  Input B is stored as uint32_t with shape: [ceil(N, 32//bits), N]:
```
Comment (yufenglee) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:3460]: >Input B is stored as uint32_t with shape: [ceil(N, 32//bits), N]: [](http://example.com/codeflow?start=2&length=65)

hqq uses different format with gptq?

File: onnxruntime/contrib_ops/cuda/quantization/matmul_nbits.cu [hunk 5]
```diff
@@ -348,6 +349,308 @@
     int shared_mem_per_block,
     cudaStream_t stream);
 
+
+namespace GPTQPacking {
+constexpr int kBlockSize = 256;
+constexpr int kNumWaves = 32;
+const int width_element_per_block = 32 * 2;
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/quantization/matmul_nbits.cu:356]: nit: following the name style for constant value and use constexpr

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc (hunk 4 above)
Comment (wejoncy) [reply onnxruntime/core/graph/contrib_ops/contrib_defs.cc:3460]: They are the same. `ceil(N*bits, 32)` is more accurate  to represent different bits like 3/5 bits.

File: docs/ContribOperators.md [hunk 6]
```diff
@@ -2808,22 +2808,14 @@ This version of the operator has been available since version 1 of the 'com.micr
...
-  
-    For a block blob. It is stored in format:
-    struct Blob {
-      uint8 one_bits[(bits & 0x1) * 1 * block_size / 8];  // highest 1 bit for 3, 5, 7 bits quantization
-      uint8 two_bits[(bits & 0x2) * 2 * block_size / 8];  // high 2 bits for 2, 6, 7 bits quantization
-      uint8 four_bits[(bits & 0x4) * 4 * block_size / 8]; // low 4 bits for 4, 5, 6 bits quantization
-    }
+    Input is stored as uint8_t with shape: [N][n_blocks_per_col][blob_size] in which:
+    - n_blocks_per_col = (K + block_size - 1) / block_size
+    - blob_size = block_size / 8 * bits
```
Comment (yufenglee) [inline docs/ContribOperators.md:2814]: could you please add a description how the data are stored?

File: onnxruntime/contrib_ops/cuda/quantization/matmul_nbits.h [hunk 7]
```diff
@@ -0,0 +1,41 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+//
+// This module define MatMulFp32Q4 operator, it is basically
```
[the remark this answers is not in the document]
Comment (yufenglee) [reply onnxruntime/contrib_ops/cuda/quantization/matmul_nbits.h:5]: and also the corresponding place in the *.cc file

File: onnxruntime/contrib_ops/cuda/quantization/matmul_nbits.cc [hunk 8]
```diff
@@ -7,9 +7,10 @@
 // pre-packed and block-compacted into int4
 //
 
-#include "core/common/safeint.h"
-#include "core/providers/cuda/cuda_kernel.h"
-#include "core/providers/cuda/shared_inc/fpgeneric.h"
+#include "contrib_ops/cuda/quantization/matmul_nbits.h"
+#include <cstdint>
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/quantization/matmul_nbits.cc:11]: nit: add a blank line before and after

File: onnxruntime/contrib_ops/cuda/quantization/matmul_nbits.cc [hunk 9]
```diff
@@ -67,33 +47,53 @@ Status MatMulNBits<T>::ComputeInternal(OpKernelContext* ctx) const {
...
+                          reinterpret_cast<const CudaT*>(scales_data),
+                          static_cast<const uint8_t*>(zero_points_data),
+                          SafeInt<int>(helper.M()),
+                          SafeInt<int>(helper.N()),
+                          SafeInt<int>(helper.K()),
+                          SafeInt<int>(block_size_),
+                          SafeInt<int>(GetDeviceProp().sharedMemPerBlock),
+                          static_cast<cudaStream_t>(ctx->GetComputeStream()->GetHandle()));
+
   if (!is_4bit_done) {
```
Comment (yufenglee) [inline onnxruntime/contrib_ops/cuda/quantization/matmul_nbits.cc:65]: nit: as the if block get larger, could you please return early if is_4bit_done is true.

File: docs/ContribOperators.md (hunk 6 above)
Comment (wejoncy) [reply docs/ContribOperators.md:2814]: Sure. Added

File: onnxruntime/contrib_ops/cuda/quantization/matmul_nbits.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (wejoncy) [reply onnxruntime/contrib_ops/cuda/quantization/matmul_nbits.h:5]: Done

**Review by yufenglee — APPROVED** [review]
:shipit:

**Pull request merged**